File: android/src/main/kotlin/com/timebox/native_llama/NativeLlamaPlugin.kt
package com.timebox.native_llama

import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.Keep
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

class NativeLlamaPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    private var eventSink: EventChannel.EventSink? = null
    private val handler = Handler(Looper.getMainLooper())
    // Executes C++ inference on a background thread to keep Flutter UI smooth
    private val executor = Executors.newSingleThreadExecutor()

    companion object {
        private const val TAG = "NativeLlamaPlugin"

        /// Whether libnative_llama.so is present and loadable on this device.
        ///
        /// The library is built for arm64-v8a only (see `abiFilters` in
        /// android/build.gradle.kts), so on any other ABI the load fails and
        /// every `external` method below would throw UnsatisfiedLinkError.
        ///
        /// This used to be a bare `System.loadLibrary` in a companion `init`,
        /// which made that an Error thrown from the *class initializer* — i.e.
        /// from `new NativeLlamaPlugin()` inside GeneratedPluginRegistrant.
        /// That registrant only catches `Exception`, and UnsatisfiedLinkError
        /// is an `Error`, so it escaped configureFlutterEngine() and killed the
        /// process before Dart main() ever ran. Google Play review (x86_64
        /// emulators) saw that as "crashes after opening", and Crashlytics
        /// never reported it because a 100%-reproducible startup crash can
        /// never upload the report it queued on the previous launch.
        ///
        /// Catching Throwable here degrades the LLM features to unavailable
        /// instead of taking the whole app down with them.
        @JvmStatic
        val isAvailable: Boolean = try {
            System.loadLibrary("native_llama")
            true
        } catch (t: Throwable) {
            Log.e(
                TAG,
                "libnative_llama.so unavailable on ABI " +
                    "${Build.SUPPORTED_ABIS.firstOrNull()} — LLM features disabled",
                t
            )
            false
        }
    }

    // Native JNI bindings
    private external fun initLlama(modelPath: String, nCtx: Int, nThreads: Int, nGpuLayers: Int): Boolean
    private external fun initDraftModel(modelPath: String, nCtx: Int, nThreads: Int, nGpuLayers: Int): Boolean

    // --- Vision/Media Model Call ---
    private external fun initVision(mmprojPath: String): Boolean

    private external fun getCpuCores(performanceOnly: Boolean): Int

    private external fun getEmbedding(text: String): DoubleArray?

    // --- MODIFIED: Renamed imagePaths to mediaPaths ---
    private external fun startNativeGeneration(roles: Array<String>, contents: Array<String>, mediaPaths: Array<String>, temperature: Float, topK: Int, topP: Float, repeatPenalty: Float, penaltyLastN: Int, freqPenalty: Float, presencePenalty: Float, grammar: String?)

    private external fun abortGeneration()

    // Recreates only the llama context (KV cache + positions), keeping weights +
    // vision projector resident — a cheap, clean reset between images.
    private external fun resetContext(nCtx: Int): Boolean

    private external fun disposeLlama()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "native_llama/methods")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "native_llama/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        // Without the native library every branch below would throw
        // UnsatisfiedLinkError off the executor thread, where nothing catches
        // it. Fail the call instead so Dart can fall back to a non-LLM path.
        if (!isAvailable) {
            result.error(
                "NATIVE_UNAVAILABLE",
                "native_llama is not available on this device's ABI " +
                    "(${Build.SUPPORTED_ABIS.firstOrNull()}).",
                null
            )
            return
        }
        when (call.method) {
            "initModel" -> {
                val modelPath = call.argument<String>("modelPath")
                val nCtx = call.argument<Int>("nCtx") ?: -1
                val nThreads = call.argument<Int>("nThreads") ?: -1
                val nGpuLayers = call.argument<Int>("nGpuLayers") ?: 0

                if (modelPath != null) {
                    executor.execute {
                        val success = initLlama(modelPath, nCtx, nThreads, nGpuLayers)
                        handler.post { result.success(success) }
                    }
                } else {
                    result.error("INVALID_ARGUMENT", "Model path is null", null)
                }
            }
            "initDraftModel" -> {
                val modelPath = call.argument<String>("modelPath")
                val nCtx = call.argument<Int>("nCtx") ?: -1
                val nThreads = call.argument<Int>("nThreads") ?: -1
                val nGpuLayers = call.argument<Int>("nGpuLayers") ?: 0

                if (modelPath != null) {
                    executor.execute {
                        val success = initDraftModel(modelPath, nCtx, nThreads, nGpuLayers)
                        handler.post { result.success(success) }
                    }
                } else {
                    result.error("INVALID_ARGUMENT", "Draft model path is null", null)
                }
            }
            "initVision" -> {
                val mmprojPath = call.argument<String>("mmprojPath")
                if (mmprojPath != null) {
                    executor.execute {
                        val success = initVision(mmprojPath)
                        handler.post { result.success(success) }
                    }
                } else {
                    result.error("INVALID_ARGUMENT", "Vision model path is null", null)
                }
            }
            "getEmbedding" -> {
                val text = call.argument<String>("text")
                if (text != null) {
                    executor.execute {
                        val embedding = getEmbedding(text)
                        handler.post {
                            if (embedding != null) result.success(embedding.toList())
                            else result.error("EMBEDDING_ERROR", "Failed to get embedding", null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENT", "Text is null", null)
                }
            }
            "startGeneration" -> {
                val roles = call.argument<List<String>>("roles")?.toTypedArray()
                val contents = call.argument<List<String>>("contents")?.toTypedArray()

                // --- MODIFIED: Extract mediaPaths (defaults to empty array if none) ---
                val mediaPaths = call.argument<List<String>>("mediaPaths")?.toTypedArray() ?: emptyArray()

                // Cast Double from Dart to Float for Kotlin/C++ boundary
                val temperature = call.argument<Double>("temperature")?.toFloat() ?: 0.7f
                val topK = call.argument<Int>("topK") ?: 40
                val topP = call.argument<Double>("topP")?.toFloat() ?: 0.9f
                // Repetition penalty (defaults preserve the previous baked-in values).
                val repeatPenalty = call.argument<Double>("repeatPenalty")?.toFloat() ?: 1.2f
                val penaltyLastN = call.argument<Int>("penaltyLastN") ?: 128
                val freqPenalty = call.argument<Double>("freqPenalty")?.toFloat() ?: 0.1f
                val presencePenalty = call.argument<Double>("presencePenalty")?.toFloat() ?: 0.1f
                // Optional GBNF grammar; null means unconstrained sampling.
                val grammar = call.argument<String>("grammar")

                if (roles != null && contents != null) {
                    executor.execute {
                        startNativeGeneration(roles, contents, mediaPaths, temperature, topK, topP, repeatPenalty, penaltyLastN, freqPenalty, presencePenalty, grammar)
                    }
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Roles or contents are null", null)
                }
            }
            "abortGeneration" -> {
                abortGeneration()
                result.success(true)
            }
            "resetContext" -> {
                val nCtx = call.argument<Int>("nCtx") ?: -1
                executor.execute {
                    val success = resetContext(nCtx)
                    handler.post { result.success(success) }
                }
            }
            "dispose" -> {
                executor.execute {
                    disposeLlama()
                    handler.post { result.success(true) }
                }
            }
            "getCpuCores" -> {
                val performanceOnly = call.argument<Boolean>("performanceOnly") ?: false
                executor.execute {
                    val cores = getCpuCores(performanceOnly)
                    handler.post { result.success(cores) }
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        // Guarded for the same reason as onMethodCall: with no native library
        // this is an UnsatisfiedLinkError on the way *out* of the engine, which
        // would crash teardown on every non-arm64 device.
        if (isAvailable) disposeLlama()
    }

    // FIX: Push EOS string explicitly so Dart can catch it manually
    @Keep
    fun onTokenReceived(token: String) {
        handler.post {
            eventSink?.success(token)
            if (token == "__END_OF_STREAM__") {
                eventSink?.endOfStream()
            }
        }
    }
}

File: ios/Classes/NativeLlamaPlugin.swift
import Flutter
import UIKit

public class NativeLlamaPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: "native_llama/methods", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "native_llama/events", binaryMessenger: registrar.messenger())

        let instance = NativeLlamaPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initModel":
            guard let args = call.arguments as? [String: Any],
                  let modelPath = args["modelPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Model path is null", details: nil))
                return
            }

            let nCtx = args["nCtx"] as? Int ?? -1
            let nThreads = args["nThreads"] as? Int ?? -1
            let nGpuLayers = args["nGpuLayers"] as? Int ?? 99

            // Move heavy initialization to a background thread
            DispatchQueue.global(qos: .userInitiated).async {
                let success = LlamaBridge.shared().initModel(modelPath, nCtx: Int32(nCtx), nThreads: Int32(nThreads), nGpuLayers: Int32(nGpuLayers))
                DispatchQueue.main.async { result(success) }
            }

        case "initDraftModel":
            guard let args = call.arguments as? [String: Any],
                  let modelPath = args["modelPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Model path is null", details: nil))
                return
            }

            let nCtx = args["nCtx"] as? Int ?? -1
            let nThreads = args["nThreads"] as? Int ?? -1
            let nGpuLayers = args["nGpuLayers"] as? Int ?? 99

            DispatchQueue.global(qos: .userInitiated).async {
                let success = LlamaBridge.shared().initDraftModel(modelPath, nCtx: Int32(nCtx), nThreads: Int32(nThreads), nGpuLayers: Int32(nGpuLayers))
                DispatchQueue.main.async { result(success) }
            }

        case "initVision":
            guard let args = call.arguments as? [String: Any],
                  let mmprojPath = args["mmprojPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Vision model path is null", details: nil))
                return
            }

            DispatchQueue.global(qos: .userInitiated).async {
                let success = LlamaBridge.shared().initVision(mmprojPath)
                DispatchQueue.main.async { result(success) }
            }

        case "getEmbedding":
            guard let args = call.arguments as? [String: Any],
                  let text = args["text"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Text is null", details: nil))
                return
            }
            DispatchQueue.global(qos: .userInitiated).async {
                let embedding = LlamaBridge.shared().getEmbedding(text)
                DispatchQueue.main.async { result(embedding) }
            }

        case "startGeneration":
            guard let args = call.arguments as? [String: Any],
                  let roles = args["roles"] as? [String],
                  let contents = args["contents"] as? [String] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Roles or contents are null", details: nil))
                return
            }

            // --- MODIFIED: Extract media paths instead of image paths ---
            let mediaPaths = args["mediaPaths"] as? [String] ?? []

            let temperature = (args["temperature"] as? NSNumber)?.floatValue ?? 0.7
            let topK = args["topK"] as? Int ?? 40
            let topP = (args["topP"] as? NSNumber)?.floatValue ?? 0.9

            // Prevent screen from sleeping during long generations
            UIApplication.shared.isIdleTimerDisabled = true

            DispatchQueue.global(qos: .userInitiated).async {
                // Pass mediaPaths to the bridge
                LlamaBridge.shared().startGeneration(withRoles: roles,
                                                    contents: contents,
                                                  mediaPaths: mediaPaths,
                                                 temperature: temperature,
                                                        topK: Int32(topK),
                                                        topP: topP) { [weak self] token in
                    guard let token = token else { return }
                    DispatchQueue.main.async {
                        if token == "__END_OF_STREAM__" {
                            UIApplication.shared.isIdleTimerDisabled = false
                            self?.eventSink?(token)
                            self?.eventSink?(FlutterEndOfEventStream)
                        } else {
                            self?.eventSink?(token)
                        }
                    }
                }
            }
            result(nil)

        case "abortGeneration":
            LlamaBridge.shared().abortGeneration()
            result(true)

        case "dispose":
            DispatchQueue.global(qos: .userInitiated).async {
                LlamaBridge.shared().dispose()
                DispatchQueue.main.async { result(true) }
            }

        case "getCpuCores":
            let performanceOnly = (call.arguments as? [String: Any])?["performanceOnly"] as? Bool ?? false
            let cores = LlamaBridge.shared().getCpuCores(performanceOnly)
            result(Int(cores))

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}